### PR TITLE
[VL] Fix potential double unload of JNI libraries on Java 11+

### PR DESCRIPTION
**Pull request by yangzhg**
## What changes were proposed in this pull request?

Previously, the addHookForLibUnloading method added a shutdown hook unconditionally to unload JNI libraries. This could potentially lead to double unload issues on Java 11 and later versions where the JVM automatically handles JNI library cleanup.

This commit modifies addHookForLibUnloading to conditionally add the unload hook based on the JVM version:
- For Java 11 and later, the method adds a no-op hook since the JVM handles the cleanup automatically.
- For Java versions prior to 11, the method adds the provided hook to ensure proper cleanup.

This change ensures compatibility and prevents potential double unload issues on Java 11 and later.

Fixes: #6145

### CODE REVIEW

**Comment by zhztheplayer** [comment]
Which library Velox's lib loader tries to unload? I remember that option (`boolean requireUnload`) was only used by CH backend.

**Comment by yangzhg** [comment]
> Which library Velox's lib loader tries to unload? I remember that option (`boolean requireUnload`) was only used by CH backend.

I'm not sure about which libs tries to unload, but I found after Java 11 there is no need to unload JNI libs, and  the unload may cause core dump or UBs.

**Comment by zhztheplayer** [comment]
> > Which library Velox's lib loader tries to unload? I remember that option (`boolean requireUnload`) was only used by CH backend.
> 
> I'm not sure about which libs tries to unload, but I found after Java 11 there is no need to unload JNI libs, and the unload may cause core dump or UBs.

Then I suppose this is not related to https://github.com/apache/incubator-gluten/issues/6145 ? I see `Fixes: #6145` in PR description.

@zzcclp Regarding lib-unloading, why CH backend needs to unload a library?

**Comment by zzcclp** [comment]
> requireUnload

There are some resources need to be released when shutting jvm, otherwise ch will core dump

**Comment by yangzhg** [comment]
> Then I suppose this is not related to #6145 ? I see `Fixes: #6145` in PR description.

apply this patch  the core dump is fixed

**Comment by yangzhg** [comment]
> resources

I will try to fix only in veloxbackend

**Comment by yangzhg** [comment]
@zzcclp @zhztheplayer I have made it only worked on Velox backend

**Comment by lwz9103** [comment]
Run Gluten Clickhouse CI

File: gluten-core/src/main/scala/org/apache/spark/util/GlutenShutdownManager.scala [hunk 1]
```diff
@@ -16,14 +16,28 @@
...
 
   def addHook(hook: () => Unit): AnyRef = {
     ShutdownHookManager.addShutdownHook(ShutdownHookManager.DEFAULT_SHUTDOWN_PRIORITY)(hook)
   }
 
   def addHookForLibUnloading(hook: () => Unit): AnyRef = {
-    ShutdownHookManager.addShutdownHook(ShutdownHookManager.SPARK_CONTEXT_SHUTDOWN_PRIORITY)(hook)
+    ShutdownHookManager.addShutdownHook(ShutdownHookManager.SPARK_CONTEXT_SHUTDOWN_PRIORITY) {
+      if (
+        BackendsApiManager.getBackendName == "velox" && SystemUtils.isJavaVersionAtLeast(
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/org/apache/spark/util/GlutenShutdownManager.scala:32]: `if (BackendsApiManager.getBackendName == "...")`

Seems like a writing we should avoid in core module's code. Let's figure out if there is a better way.

**Review by zhztheplayer — COMMENTED** [review]
Thanks. But let's find another way rather than checking backend type in core module.

**Comment by zhztheplayer** [comment]
@yangzhg I could not reproduce the unloading error since the static storage `REQUIRE_UNLOAD_LIBRARY_PATHS` looks to be always empty in Velox backend thus the shutdown hook should ideally do nothing.

image

For different build configurations I assume this list should either be empty since we always pass `requireUnload=false` when using JniLibLoader in Velox backend. Maybe I miss something?

Would you like to collect some log or debug info to see which library is being unloaded in your case? Let's find out the root cause prior to doing any code changes. Thank you very much.

**Pull request closed without merging**